objects/src/accounts/vault.rs

```diff
@@ -0,0 +1,68 @@
+use super::{AccountError, AccountId, Asset, Digest};
+
+// ACCOUNT VAULT
+// ================================================================================================
+
+/// An asset container for an account.
+///
+/// An account vault can contain an unlimited number of assets. The assets are stored in a Sparse
+/// Merkle tree as follows:
+/// - For fungible assets, the index of a node is defined by the issuing faucet ID, and the value
+///   of the node is the asset itself. Thus, for any fungible asset there will be only one node
+///   in the tree.
+/// - For non-fungible assets, the index is defined by the asset itself, and the asset is also
+///   the value of the node.
+///
+/// An account vault can be reduced to a single hash which is the root of the Sparse Merkle tree.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct AccountVault {
+    // TODO: add backing sparse Merkle tree
+    assets: Vec<Asset>,
+}
+
+impl AccountVault {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new account vault initialized with the provided assets.
+    ///
+    /// TODO: return error if there are duplicates in the provided asset list.
+    pub fn new(assets: &[Asset]) -> Self {
+        Self {
+            // TODO: put assets into a Sparse Merkle trees
+            assets: assets.to_vec(),
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a commitment to this vault.
+    pub fn root(&self) -> Digest {
+        todo!()
+    }
+
+    /// Returns true if the specified non-fungible asset is stored in this vault.
+    pub fn has_nfasset(&self, asset: Asset) -> Result<bool, AccountError> {
```

Comment:
we probably shouldn't take the asset as owned here as we are going to test it's existence in the set. likely, we will just derive some key from it and check; this means we should either ask for a reference of the key, or a reference to the asset so we can compute the key.

Reply:
For non-fungible assets, the assets itself is the key. We may change this later on (i.e., keep asset as the key, but store something like insertion time as the value) - but that requires more thinking.

As for taking it as owned: `Asset` implements `Copy` - so, and I usually try to pass `Copy` structs by value. But maybe that's not the best way to do it?